File: src/library/patterns/Firework.ts
import Simulator from '../Simulator'
import SinglePattern from './SinglePattern'
import Color from '../Color'

export default class Snake extends SinglePattern {

	row: number
	col: number
	maxRadius: number = 5
	currentRadius: number = 1

	constructor(simulator: Simulator, row: number, col: number) {
		super(simulator);
		this.row = row;
		this.col = col;
	}

	tick() {
		if (this.currentRadius >= this.maxRadius) {
			console.log("DONE")
			return false;
		}
		this.simulator.setLED(this.row, this.col, new Color(100, 200, 50))
		
		for (let i = 1; i < this.currentRadius; i++) {
			// Q1
			this.simulator.setLED(this.row - i, this.col + i, new Color(100, 200, 50))
			// Q2
			this.simulator.setLED(this.row - i, this.col - i, new Color(100, 200, 50))
			// Q3
			this.simulator.setLED(this.row + i, this.col - i, new Color(100, 200, 50))
			// Q4
			this.simulator.setLED(this.row + i, this.col + i, new Color(100, 200, 50))
		}
		this.currentRadius++;
		return true
	}
}

File: cpp/src/Firework.cpp
#include "Firework.h"
#include "Controller.h"
#include "Color.h"

bool Firework::tick() {
  if (currentRadius_ >= maxRadius_) {
    return false;
  }

  controller_.setLED(x_, y_, {100, 200, 50});

  for (int i = 1; i < currentRadius_; i++) {
    // Q1
    controller_.setLED(x_ - i, y_ + i, {100, 200, 50});
    // Q2
    controller_.setLED(x_ - i, y_ - i, {100, 200, 50});
    // Q3
    controller_.setLED(x_ + i, y_ - i, {100, 200, 50});
    // Q4
    controller_.setLED(x_ + i, y_ + i, {100, 200, 50});
  }
  currentRadius_++;
  return true;
}
